mongodb-atlas/cmd/main.go

```diff
@@ -0,0 +1,253 @@
+	orgID       string
+	rateLimiter *rate.Limiter
+	atlasClient HTTPClient
+}
+
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+func (a *AtlasCostSource) GetCustomCosts(req *pb.CustomCostRequest) []*pb.CustomCostResponse {
+	results := []*pb.CustomCostResponse{}
+
+	targets, err := opencost.GetWindows(req.Start.AsTime(), req.End.AsTime(), req.Resolution.AsDuration())
+	if err != nil {
+		log.Errorf("error getting windows: %v", err)
+		errResp := pb.CustomCostResponse{
+			Errors: []string{fmt.Sprintf("error getting windows: %v", err)},
+		}
+		results = append(results, &errResp)
+		return results
+	}
+
+	for _, target := range targets {
+		if target.Start().After(time.Now().UTC()) {
+			log.Debugf("skipping future window %v", target)
+			continue
+		}
+
+		log.Debugf("fetching atlas costs for window %v", target)
+		result, err := a.getAtlasCostsForWindow(&target)
+		if err != nil {
+			log.Errorf("error getting costs for window %v: %v", target, err)
+			errResp := pb.CustomCostResponse{}
+			errResp.Errors = append(errResp.Errors, fmt.Sprintf("error getting costs for window %v: %v", target, err))
+			results = append(results, &errResp)
+		} else {
+			results = append(results, result)
+		}
+	}
+
+	return results
+}
+
+func (a *AtlasCostSource) getAtlasCostsForWindow(win *opencost.Window) (*pb.CustomCostResponse, error) {
+
+	// get the token
+	token, err := CreateCostExplorerQueryToken(a.orgID, *win.Start(), *win.End(), a.atlasClient)
+	if err != nil {
+		log.Errorf("error getting token: %v", err)
+		return nil, err
+	}
+
+	// get the costs
+	costs, err := GetCosts(a.atlasClient, a.orgID, token)
+	if err != nil {
+		log.Errorf("error getting costs: %v", err)
+		return nil, err
+	}
+
+	resp := pb.CustomCostResponse{
+		Metadata:   map[string]string{"api_client_version": "v1"},
+		CostSource: "data_storage",
+		Domain:     "mongodb-atlas",
+		Version:    "v1",
+		Currency:   "USD",
+		Start:      timestamppb.New(*win.Start()),
+		End:        timestamppb.New(*win.End()),
+		Errors:     []string{},
+		Costs:      costs,
+	}
+	return &resp, nil
+}
+
+func GetCosts(client HTTPClient, org string, token string) ([]*pb.CustomCost, error) {
+	request, _ := http.NewRequest("GET", fmt.Sprintf(costExplorerQueryFmt, org, token), nil)
+
+	request.Header.Set("Accept", "application/vnd.atlas.2023-01-01+json")
+	request.Header.Set("Content-Type", "application/vnd.atlas.2023-01-01+json")
+
+	response, error := client.Do(request)
+	statusCode := response.StatusCode
+	//102 status code means processing - so repeat call 2 times to see if we get a response
+	for count := 1; count < 2 && statusCode == http.StatusProcessing; count++ {
+		// Sleep for 5 seconds before the next request
+		time.Sleep(5 * time.Second)
+		response, _ := client.Do(request)
+		statusCode = response.StatusCode
+
+	}
+
+	if statusCode == http.StatusProcessing {
+		msg := "timeout waiting for response"
+		return nil, fmt.Errorf(msg)
+	}
+	if error != nil {
```

Comment:
i would be very very very careful about using error as a variable name. It is a protected name in many situations in go since it is the type of the error values. 95% of the time, we use err as the name to avoid confusion